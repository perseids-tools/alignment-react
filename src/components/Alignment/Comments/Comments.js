import React from 'react';

import SentenceContext from '../sentence-context';

import Comments from '../../Comments';

const CommentsWithContext = () => (
  <SentenceContext.Consumer>
    {({ sentence, active }) => (
      <Comments
        sentence={sentence}
        active={active}
      />
    )}
  </SentenceContext.Consumer>
);

export default CommentsWithContext;
