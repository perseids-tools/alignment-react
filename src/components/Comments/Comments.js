import React from 'react';

import { activeType, sentenceType } from '../../types';

import styles from './Comments.module.scss';

// eslint-disable-next-line react/prop-types
const commentDl = ({ title, value }, lnum, n, ii) => (
  <dl key={`comment-${lnum}-${n}-${ii}`} className={styles.dl}>
    <div className={styles.container}>
      <dt className={styles.dt}>{title}</dt>
      <dd className={styles.dd}>{value}</dd>
    </div>
  </dl>
);

const wordListDls = ({ wds, active }) => {
  const dls = [];

  (wds || []).forEach(({ $: { lnum }, w }) => (
    (w || []).forEach(({ $: { n }, comment, text }) => (
      (comment || []).forEach(({ _: commentText }, ii) => {
        const title = (text || []).map(({ _ }) => _).join(' ');

        if (active && active.aligned && active.aligned[lnum] && active.aligned[lnum].has(n)) {
          dls.push(commentDl({ title, value: commentText }, lnum, n, ii));
        }
      })
    ))
  ));

  return dls;
};

const Comments = ({ sentence, active }) => {
  if (!active) {
    return false;
  }

  const wds = sentence && sentence.wds;
  const dls = wordListDls({ wds, active });

  if (dls.length === 0) {
    return false;
  }

  return (
    <div className={styles.comments}>{dls}</div>
  );
};

Comments.propTypes = {
  sentence: sentenceType.isRequired,
  active: activeType,
};

Comments.defaultProps = {
  active: null,
};

export default Comments;
