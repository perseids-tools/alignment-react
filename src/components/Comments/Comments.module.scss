.comments {
  background-color: rgb(242, 242, 242);

  padding-top: 0.2rem;
  padding-bottom: 0.2rem;
  padding-left: 0.4rem;
  padding-right: 0.4rem;

  margin-bottom: 0.2rem;
}

.dl {
  display: flex;
  flex-wrap: wrap;
}

.container {
  margin-right: 7px;
  margin-left: 7px;
  margin-bottom: 2px;
  margin-top: 2px;
}

.dt {
  font-weight: bold;
  margin: 0;
  padding: 0;
}

.dd {
  margin: 0;
  padding: 0;
}

