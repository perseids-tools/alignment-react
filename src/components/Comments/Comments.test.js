import React from 'react';
import renderer from 'react-test-renderer';

import Comments from './Comments';

const sentence = {
  $: { n: '1' },
  wds: [
    {
      $: { lnum: 'L1' },
      w: [
        { $: { n: '1-1' }, text: [{ _: 'A' }], refs: [{ $: { nrefs: '1-1' } }] },
        { $: { n: '1-2' }, text: [{ _: 'BC' }], refs: [{ $: { nrefs: '1-2 1-3' } }] },
        {
          $: { n: '1-3' },
          text: [{ _: 'D' }],
          comment: [{ _: 'Test', $: { class: 'mark' } }],
          refs: [{ $: { nrefs: '1-3 1-4' } }],
        },
        { $: { n: '1-4' }, text: [{ _: 'E' }], refs: [{ $: { nrefs: '1-4' } }] },
        { $: { n: '1-5' }, text: [{ _: 'F' }], refs: [{ $: { nrefs: '1-5 1-6' } }] },
        { $: { n: '1-6' }, text: [{ _: 'G' }], refs: [{ $: { nrefs: '1-5 1-6' } }] },
      ],
    },
    {
      $: { lnum: 'L2' },
      w: [
        { $: { n: '1-1' }, text: [{ _: 'A' }], refs: [{ $: { nrefs: '1-1' } }] },
        { $: { n: '1-2' }, text: [{ _: 'B' }], refs: [{ $: { nrefs: '1-2' } }] },
        {
          $: { n: '1-3' },
          text: [{ _: 'C' }],
          comment: [{ _: 'Other languge test', $: { class: 'mark' } }],
          refs: [{ $: { nrefs: '1-3' } }],
        },
        { $: { n: '1-4' }, text: [{ _: 'DE' }], refs: [{ $: { nrefs: '1-3' } }] },
        { $: { n: '1-5' }, text: [{ _: 'F' }], refs: [{ $: { nrefs: '1-5 1-6' } }] },
        { $: { n: '1-6' }, text: [{ _: 'G' }], refs: [{ $: { nrefs: '1-5 1-6' } }] },
      ],
    },
  ],
};

it('does not render when there is no active', () => {
  const component = (
    <Comments sentence={sentence} />
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});

it('does not render when the active has no comment', () => {
  const active = {
    selected: { lnum: 'L1', n: '1-1' },
    aligned: {
      L1: new Set(['1-1']),
      L2: new Set(['1-1']),
    },
  };

  const component = (
    <Comments
      sentence={sentence}
      active={active}
    />
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});

it('displays all comments associated with active words', () => {
  const active = {
    selected: { lnum: 'L1', n: '1-1' },
    aligned: {
      L1: new Set(['1-3']),
      L2: new Set(['1-3', '1-4']),
    },
  };

  const component = (
    <Comments
      sentence={sentence}
      active={active}
    />
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});
