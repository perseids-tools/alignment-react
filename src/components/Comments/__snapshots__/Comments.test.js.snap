// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays all comments associated with active words 1`] = `
<div
  className="comments"
>
  <dl
    className="dl"
  >
    <div
      className="container"
    >
      <dt
        className="dt"
      >
        D
      </dt>
      <dd
        className="dd"
      >
        Test
      </dd>
    </div>
  </dl>
  <dl
    className="dl"
  >
    <div
      className="container"
    >
      <dt
        className="dt"
      >
        C
      </dt>
      <dd
        className="dd"
      >
        Other languge test
      </dd>
    </div>
  </dl>
</div>
`;

exports[`does not render when the active has no comment 1`] = `null`;

exports[`does not render when there is no active 1`] = `null`;
